Add CMake support for building and linking against CaDiCaL using the IPASIR interface

NlightNFotis opened the pull request:
Expand the CMake build to be able to link against IPASIR supporting solvers. Two
new solver configurations have been added:

* CaDiCaL, which should be easy to compile against (`-Dsat_impl=ipasir_cadical` will handle downloading, building and linking against the solver), and
* Any arbitrary solver supporting the IPASIR interface (for the more ~adventurous~ demanding users, who may want to link against a different variety of solvers). This is more involved as an option (requires a number of different flags to be passed to CMake).

This expands our configuration options significantly. Previously, we could do one of two things:
  
  * Build with `cmake` using the internal CaDiCaL interface (`satcheck_cadical`), or
  * Build with `make` using the IPASIR interface.

This now allows us to build using both of the build systems, and link with a number of different solvers, for both testing and production use.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Add CMake support for building CBMC with IPASIR linking against CaDiCaL.

NlightNFotis committed:
Add compile instructions for CaDiCaL-IPASIR through cmake.

TGWDB commented on src/solvers/CMakeLists.txt:
```diff
@@ -51,6 +55,7 @@ list(REMOVE_ITEM sources
     ${lingeling_source}
     ${booleforce_source}
     ${minibdd_source}
+    # ${ipasir_source}
```
Commented out option?

NlightNFotis replied:
No - this is a weird mechanism in the cmake file, where the files specified there are _removed_ from the build, and then re-introduced later, when the particular solver to be linked against has been identified.

I haven't understood the mechanism exactly (how it reintroduces the files for, say, minisat or glucose) so I've left this there as a reminder for me to investigate further. If I remove the commenting out of the ipasir sources there, what happens is that we get a linking error when building `smt2_solver`, where it complains that it can't find references for `satcheck_ipasir` (i.e. failing to build `satcheck_ipasir.cpp`).

TGWDB commented on src/solvers/CMakeLists.txt:
```diff
@@ -128,6 +133,33 @@ elseif("${sat_impl}" STREQUAL "cadical")
+    message(STATUS "Building with IPASIR solver linking against: CaDiCaL")
+
+    download_project(PROJ cadical
+        URL https://github.com/arminbiere/cadical/archive/rel-1.4.0.tar.gz
```
This uses CaDiCaL 1.4.0 and the non-IPASIR path uses CaDiCaL 1.3.0, can these by made the same version?

NlightNFotis replied:
Yes, I have it on my todo list, though it needs to be submitted as a different PR - I prefer PRs to be isolated to single feature changes.

tautschnig replied:
I'd have the commit accomplishing this at https://github.com/tautschnig/cbmc/commit/330542a37c768f00f71959b2b2229336f15c3079, but thus far held back on creating a PR as performance needs to be evaluated (I vaguely recall a regression).

TGWDB commented:
I'm curious how much of the IPASIR building configuration is specific to CaDiCaL and how much is generic? Is it easy to link against another IPASIR compatible solver if one already has the correct path information to pass in to `cmake`? If this is a small addition/extension, this might be nice to include here.

NlightNFotis committed:
Add custom SAT-solver support through IPASIR to CMAKE.

This is allowing an advanced user to build and link CBMC
with arbitrary SAT Solvers using the IPASIR interface.

NlightNFotis committed:
Add instructions on how to link against IPASIR solvers through CMAKE.

Add more instructions on linking against custom IPASIR
supporting solvers through CMake.

TGWDB commented on COMPILING.md:
````diff
@@ -462,6 +462,27 @@ Note that at the time of writing this has been tested to work with the CaDiCaL
+controlled with the flag `-Dsat_impl=ipasir-cadical`, like so:
+
+```sh
+$ cmake -Bbuild_ipasir -S . -Dsat-impl=ipasir-cadical
````
`-Dsat_impl`

TGWDB commented on COMPILING.md:
````diff
@@ -462,6 +462,27 @@ Note that at the time of writing this has been tested to work with the CaDiCaL
+an invocation like this is needed:
+
+```sh
+$ cmake -Bbuild_ipasir -S . -Dsat-impl=ipasir-custom -DIPASIR=<source_location> -DIPASIR_LIB=<lib_location>
````
`-Dsat_impl`

TGWDB commented on COMPILING.md:
````diff
@@ -462,6 +462,27 @@ Note that at the time of writing this has been tested to work with the CaDiCaL
+$ cmake -Bbuild_ipasir -S . -Dsat-impl=ipasir-custom -DIPASIR=<source_location> -DIPASIR_LIB=<lib_location>
+```
+
+with `<source_location>` being the path to the folder containing the solver implementation
+and `<lib_location>` being the path that contains a precompiled static library of
+the solver (`.a` file).
````
Note that these paths must be absolute and not relative!

TGWDB commented:
I've checked this on Ubuntu 20.04 with the CaDiCaL and also linked to SATCOMP via the IPASIR path, all worked for me.

chrisr-diffblue reviewed (approved):
Looks good to me. Thanks for the clear instructions!
